Feature: Custom Memory Allocator

suhasravi7 opened the pull request:
## Summary
Adds custom memory allocator support to the JOSE library, allowing applications to override default malloc/realloc/free functions for specialized memory management needs.

## Changes
- Added `jose_set_alloc()` and `jose_get_alloc()` API functions
- Added comprehensive test suite (`api_mem.c`)
- Updated build system integration

## Use Cases
- Memory debugging and tracking
- Custom memory pools
- Secure memory allocation
- Cross-DLL memory management

## Testing
```bash
meson test -C build api_mem
```

All tests pass, validating functionality across different usage scenarios.

suhasravi7 commented:
> The size_t type is not a built-in keyword in C; it's defined in standard library headers like <stdlib.h>. Please, fix `include/jose/cfg.h` to include the corresponding one

Updated the PR

sarroutbi commented:
Hello @suhasravi7 . Thanks for the patch.

I guess it is not as easy as I thought ... could you please merge [next change](https://github.com/suhasravi7/jose/pull/1) to your branch and incorporate it to the PR to check if compilation errors are fixed?

suhasravi7 commented:
> Hello @suhasravi7 : Could you please justify why this should be merged? We don't see a clear requirement to change memory allocators, and we find some security constraints on including it ... which architecture, software, product, ... justifies this change? CC @sergio-correia

Thank you for raising these concerns @sarroutbi.

The motivation for introducing a custom memory allocator support in JOSE stems from our deployment environment, where we utilize an H2O proxy(https://github.com/h2o/h2o). H2O internally pre-allocates and manages memory blocks to optimize performance and reduce allocation overhead, especially under high concurrency. To align with this architecture and fully leverage these performance benefits, we require JOSE to support custom memory allocators.

By enabling JOSE to use externally managed allocators, we can maintain consistent memory management strategies across our stack. This approach offers measurable improvements in allocation speed and reduces fragmentation, compared to standard malloc/calloc routines. Additionally, it allows for tighter control over memory usage, which is critical in high-performance and resource-constrained environments.

Regarding security, the implementation ensures that allocator hooks are strictly opt-in and can be sandboxed or audited according to deployment requirements. The change is modular and does not impact default behavior unless explicitly configured.

suhasravi7 commented:
> should the api jose_cfg_set_alloc enforce consistency in the passed parameters. either all should be NULL or Non-NULL. a combination of NULL and Non Null paramters could lead to mis management of the allocated memory here?

@vikaspushkar updated the pr to handle this case

sarroutbi reviewed (approved):
LGTM

simo5 commented on lib/cfg.c:
```diff
@@ -30,6 +30,11 @@ struct jose_cfg {
+/* Global static function pointers for allocators */
+static jose_malloc_t jose_global_malloc = NULL;
+static jose_realloc_t jose_global_realloc = NULL;
+static jose_free_t jose_global_free = NULL;
```
Is there a reason why you have separate statics and then functions that at runtime do checks in order to chose which implementation to use instead of just declaring thse statics with the new function name and assigning it the standard allocation functions by default?

IE:
`static jose_malloc_t jose_malloc = malloc;`

And then jose_set_alloc() just sets
`jose_malloc = pmalloc;`

This gets rid of the unnecessary indirection on calling these functions (except for the jose_calloc variant if you do not want to force people to provide that one obligatorily).

suhasravi7 replied:
Sure @simo5 , i have updated the pull request with the suggested changes.

simo5 replied:
Sorry @suhasravi7 but that's nothing like I asked, you just changed the name of the globals and kept everything substantially the same.

sarroutbi replied:
@suhasravi7 : I believe what Simo is asking for is:

That this line:
```bash
static jose_malloc_t g_malloc = malloc;
```
should really be:
```bash
static jose_malloc_t jose_malloc = malloc;
```
and `jose_set_alloc()`  should look like:
```
int
jose_set_alloc(jose_malloc_t pmalloc, jose_realloc_t prealloc, jose_free_t pfree)
{
    /* checks here ... */
    jose_malloc = pmalloc ...
    jose_realloc = prealloc ...
    jose_free = pfree ...
}
```
and that the functions you added that just look at the global vars and call stuff should go away entirely
OTOH, Why are you not assigning `calloc` to `jose_calloc` and instead rewriting `jose_calloc` as a wrapper that reimplements `calloc`?
You should NOT reimplement it

suhasravi7 replied:
Hello @sarroutbi and @simo5 
I have updated the pull request with the suggested changes

simo5 reviewed (commented):
Just a comment, but I think this code can be made more compact and efficient.

suhravis committed:
Add custom memory allocator support

- Implement global function pointers for memory allocation
- Add jose_set_alloc(), jose_get_alloc(), jose_reset_alloc()
- Add comprehensive test suite in api_mem.c
- Remove cfg dependencies from io.c structs
- All tests passing (26/26)

simo5 reviewed (commented):
LGTM
